### PR TITLE
DEV-99: Extend selection during auto-scroll in DragToScroll plugin

### PR DESCRIPTION
**Pull request by budnix**
### Context

While a user holds LMB on a cell and drags the mouse past the viewport edge, the DragToScroll plugin already auto-scrolls the viewport periodically. Before this PR the regular drag-select stopped at the edge cell — the viewport scrolled away under it but the selection did not follow. Autofill already worked (via its own `afterScroll` replay mechanism in the autofill plugin).

This PR extends DragToScroll to update the active selection on every auto-scroll tick so drag-select tracks the newly-revealed cells, matching the behavior of Excel, Google Sheets, and other spreadsheet tools. Drag back past the opposite edge and the selection shrinks naturally.

**What changed:**

- **Selection follows auto-scroll** — on each scroll tick `afterScroll` fires, `getCellCoordsFromMousePosition` maps the last known mouse position to the current edge cell, and `selection.setRangeEnd` extends the range to that cell.
- **Direction-aware snapping** — the horizontal scroller uses `lastFullyVisibleColumn + 1` (advance) and `firstPartiallyVisibleColumn - 1` (retreat) with Walkontable's auto-snapping so the newly-revealed column sticks to the correct viewport edge.
- **Window-scroll boundary fix** — when the window is the scroll container, viewport bounds were computed from `tableOffset` which shifts on every scroll tick, causing selection to jump. Fixed to clamp against `window.innerWidth/innerHeight` (or `Math.min(0, tableOffset)` for the RTL max-left edge case).
- **Stop-scroll-on-return** — auto-scroll stops immediately when the cursor re-enters the viewport (previously it continued until the next timeout tick).
- **Walkontable `onMouseMove` handler** — a new document-level `mousemove` listener in Walkontable detects out-of-viewport pointer positions and fires the new `onCellMouseOverOutside` internal callback, which bridges into two new public hooks.
- **New public hooks:**
  - `beforeOnCellMouseOverOutside(event, coords, TD, controller)` — fired before selection changes are applied during an outside-viewport drag tick; use `controller.row/column/cell = true` to suppress selection changes (used internally by `manualRowMove` and `manualColumnMove`).
  - `afterOnCellMouseOverOutside(event, coords, TD)` — fired after selection changes for each outside-viewport drag tick.
- **RTL support** — all four directions tested, including the RTL max-left-scroll edge case where `tableOffset.left > window.innerWidth`.
- **Robustness** — all `getCell()` → `getBoundingClientRect()` calls guarded with `instanceof HTMLElement` to prevent crashes when cells are not yet rendered during rapid scroll ticks.

Covers all four directions, RTL layout, window-scroll and element-scroll containers, fixed-column/row overlays, `manualRowMove`/`manualColumnMove` move-drag suppression, single-selection-mode guard, end-of-data axis stop, and redundant-extend dedup.

ClickUp task: https://app.clickup.com/t/DEV-99

### How has this been tested?

Tested the changes locally and covered the feature with new tests.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):

1. ClickUp task DEV-99

### Affected project(s):

- [x] `handsontable`
- [ ] `@handsontable/angular-wrapper`
- [ ] `@handsontable/react-wrapper`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [x] My change requires a change to the documentation.

---

> [!NOTE]
> **Medium Risk**
> Touches core mouse/selection/scroll handling (including new hooks) and could affect drag selection/autofill and RTL/window-scrolling edge cases, though changes are well-covered by new unit/spec tests and docs.
> 
> **Overview**
> Dragging a selection beyond the viewport edge now *continues extending the selection while the grid auto-scrolls*, matching spreadsheet behavior.
> 
> This adds document-level mouse tracking in Walkontable to detect out-of-viewport drags and emit new hooks (`beforeOnCellMouseOverOutside`/`afterOnCellMouseOverOutside`), plus a new `getCellCoordsFromMousePosition` helper (with RTL/window-scroll boundary fixes) to reliably map the pointer to the current edge cell during scrolling. Selection range expansion is updated to optionally preserve direction, overlay scroll setters are hardened for window vs element scrolling, and autofill/fill-handle logic and tests are adjusted accordingly.
> 
> Documentation is added for `dragToScroll` (including speed configuration) with new JS/React/Angular examples and an interval chart, and a changelog entry announces the behavior change.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 13ddb84d257bf6e6b54750830836ffab627a9cee. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qunabu** [comment]
Task linked: [DEV-99 The selection movement doesn't scroll the viewport anymore](https://app.clickup.com/t/9015210959/DEV-99)

File: handsontable/src/helpers/dom/cellCoords.js [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+      if (scrollRelativeY < 0) {
+        foundRow = firstPartiallyVisibleRow ?? 0;
+      } else {
+        foundRow = lastPartiallyVisibleRow ?? (hotInstance.countRows() - 1);
+      }
+    }
+  }
+
+  return hotInstance._createCellCoords(foundRow, foundColumn);
+}
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/helpers/dom/cellCoords.js:272]: The HOT version handles hidden rows/columns via _IndexMapper_ and supports colspan/rowspan, while the Walkontable version operates on renderable indexes directly. They share the same structure but can't be trivially unified without crossing the Walkontable architecture boundary. The duplication is acceptable given the constraints.

File: handsontable/src/3rdparty/walkontable/src/event.js [hunk 2]
```diff
@@ -273,6 +283,167 @@ class Event {
...
+      }
+    }
+
+    return {
+      coords: wot.createCellCoords(foundRow, foundColumn),
+      isOutside: mouseX < tableViewportLeft ||
+                 mouseX > tableViewportRight ||
+                 mouseY < tableViewportTop ||
+                 mouseY > tableViewportBottom,
+    };
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/3rdparty/walkontable/src/event.js:489]: The HOT version handles hidden rows/columns via IndexMapper and supports colspan/rowspan, while the Walkontable version operates on renderable indexes directly. They share the same structure but can't be trivially unified without crossing the Walkontable architecture boundary. The duplication is acceptable given the constraints.

**Review by qunabu — DISMISSED** [review]

**Comment by adrianspdev** [comment]
@budnix list of potential fixes and improvements below:

  1. Remove leftover public field shadows in walkontable/src/event.js
  - handsontable/src/3rdparty/walkontable/src/event.js:83-92
  - The constructor now assigns wtSettings, wtTable, parent, and facadeGetter to both #-private and public fields. No reader of the public fields exists in the diff — looks like leftover scaffolding. Violates the #-private rule and accidentally enlarges the public API
  surface.
  - Fix: delete the four public assignments; keep only the #-prefixed ones.

  2. Deduplicate getCellCoordsFromMousePosition
  - handsontable/src/3rdparty/walkontable/src/event.js:115-347 re-implements handsontable/src/helpers/dom/cellCoords.js:1-243.
  - The two copies have already drifted: the Walkontable copy has commented-out colspan/rowspan handling, hard-codes firstFixedColumn = 0 (skipping IndexMapper.getVisualFromRenderableIndex, so it mis-fires on grids with the first column hidden), and lacks the
  off-screen-table fallbacks (?? hot.countCols() - 1).
  - Fix: delete the nested helpers from event.js and call the core getCellCoordsFromMousePosition (via the facade), or make the core helper a thin wrapper around a Walkontable-level helper. One source of truth.

  3. Re-evaluate mousemove → mouseOver registration
  - handsontable/src/selection/mouseEventHandler.js:157
  - Every mousemove inside the table now goes through the full mouseover/selection-extension pipeline. Risks double-firing with the existing mouseover event during drag, retriggering selection.setRangeEnd on stationary touchpad-jitter mousemoves, and interacting poorly
  with the new Walkontable document-level mousemove listener.
  - Fix: dedupe by tracking the last (row, col) handled and skipping mousemove when the cell hasn't changed; add a regression test that asserts beforeOnCellMouseOver does not double-fire.

**Comment by budnix** [comment]
> 1. Remove leftover public field shadows in walkontable/src/event.js
> 
> * handsontable/src/3rdparty/walkontable/src/event.js:83-92
> * The constructor now assigns wtSettings, wtTable, parent, and facadeGetter to both #-private and public fields. No reader of the public fields exists in the diff — looks like leftover scaffolding. Violates the #-private rule and accidentally enlarges the public API
>   surface.
> * Fix: delete the four public assignments; keep only the #-prefixed ones.

I did not modified that file range 🤔 All is #-prefixed
image

**Comment by budnix** [comment]
> 3\. Deduplicate getCellCoordsFromMousePosition

I decided to keep it split because the HOT version handles hidden rows/columns via IndexMapper and supports colspan/rowspan, while the Walkontable version operates on renderable indexes directly. They share the same structure but can't be trivially unified without crossing the Walkontable architecture boundary. The duplication is acceptable given these constraints (https://github.com/handsontable/handsontable/pull/12344#discussion_r3152173739).

**Comment by budnix** [comment]
> 3\. Re-evaluate mousemove → mouseOver registration

Fixed within https://github.com/handsontable/handsontable/pull/12344/changes/f132e1f3c30875cd8f1831d0bbaa1b38767dd523

File: handsontable/src/3rdparty/walkontable/src/cell/range.js [hunk 3]
```diff
@@ -348,19 +348,42 @@ class CellRange {
...
+        }
+
+        if (cellCoords.col < topStart.col && horizontalDirection === 'W-E') {
+          this.flipDirectionHorizontally();
+
+        } else if (cellCoords.col > bottomEnd.col && horizontalDirection === 'E-W') {
+          this.flipDirectionHorizontally();
+        }
+      }
+
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/3rdparty/walkontable/src/cell/range.js:386]: The `expand(coords, changeDirection = true)` default is intentionally non-breaking because the flip conditions are guarded by the range's pre-existing direction (`S-N`/`E-W`), so a fresh single-cell drag — which always resolves to `NW-SE`/`N-S`/`W-E` — never triggers the flip, leaving `Selection.add()` and `visualSelection.js` behaviorally identical to before for all normal drag-to-scroll paths.

File: handsontable/src/plugins/autofill/__tests__/helpers/utils.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Get the fill handle border range.
+ *
+ * @returns {CellRange}
+ */
+export function getFillHandleBorderRange() {
+  return selection().highlight.getFill().visualCellRange;
+}
```
[the remark this answers is not in the document]
Comment (budnix) [reply handsontable/src/plugins/autofill/__tests__/helpers/utils.js:8]: `getFillHandleBorderRange` is a globally available E2E test helper imported at the spec suite level, so it does not need to appear as an explicit import in individual test files.

**Review by adrianspdev — APPROVED** [review]
